libc/socket/unix-socket.c

```diff
@@ -904,6 +933,399 @@ TEST(test_unix_socket, poll)
 }
 
 
+static void read_msg(int fd, pid_t pid, int send_flags)
+{
+	int msg_len = 128;
+	int rv;
+
+	memset(buf, 0, msg_len);
+	if ((rv = read(fd, buf, msg_len)) < 0)
+		FAIL_OR_EXIT(pid, "read");
+
+	if (msg_len != rv)
+		FAIL_OR_EXIT(pid, "msg_len != rv");
+
+	if (0 != strncmp(buf, data, msg_len))
+		FAIL_OR_EXIT(pid, "strncmp != 0");
+}
+
+
+static void send_msg(int fd, pid_t pid, int send_flags)
+{
+	int msg_len = 128;
+	int rv;
+
+	if ((rv = send(fd, data, msg_len, send_flags)) < 0)
+		FAIL_OR_EXIT(pid, "read");
+
+	if (msg_len != rv)
+		FAIL_OR_EXIT(pid, "msg_len != rv");
+}
+
+
+static void unix_accept_connect_errnos(int type)
+{
+	int fd, named, rv, conn;
+
+	const char *socket_name = "/tmp/test_accept_connect_errnos";
+
+	rv = connect_to_named(BAD_FD, socket_name);
+	TEST_ASSERT(rv < 0);
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	if ((fd = socket(AF_UNIX, type, PF_UNSPEC)) < 0)
+		FAIL("socket");
+
+	rv = connect_to_named(fd, socket_name);
+	TEST_ASSERT(rv < 0);
+	TEST_ASSERT_EQUAL_INT(ECONNREFUSED, errno);
+
+	if ((named = unix_named_socket(type, socket_name)) < 0)
+		FAIL("unix_named_socket");
+
+	if (set_nonblock(named, 1) < 0)
+		FAIL("set_nonblock");
+
+	if (listen(named, 0) < 0)
+		FAIL("listen");
+
+	conn = accept(named, NULL, NULL);
+	TEST_ASSERT(conn < 0);
+	TEST_ASSERT_EQUAL_INT(EWOULDBLOCK, errno);
+
+	if (set_nonblock(fd, 1) < 0)
+		FAIL("set_nonblock");
+
+	rv = connect_to_named(fd, socket_name);
+	TEST_ASSERT(rv < 0);
+	TEST_ASSERT_EQUAL_INT(EINPROGRESS, errno);
+
+	rv = connect_to_named(fd, socket_name);
+	TEST_ASSERT(rv < 0);
+	TEST_ASSERT_EQUAL_INT(EALREADY, errno);
+
```

Comment:
Could add EISCONN test case?

Reply:
Added EISCONN check to async connection test